File: src/test/java/com/digitalpetri/fsm/dsl/GuardBuilderTest.java
/*
 * Copyright (c) 2024 Kevin Herron
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package com.digitalpetri.fsm.dsl;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.digitalpetri.fsm.FsmContext;
import java.util.LinkedList;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Predicate;
import org.junit.jupiter.api.Test;

class GuardBuilderTest {

  @Test
  void guardBuilder() {
    var transition = new PredicatedTransition<State, Event>(
        s -> s == State.S1,
        e -> e instanceof Event.E1,
        State.S2
    );

    var actions = new LinkedList<TransitionAction<State, Event>>();
    var guardBuilder = new GuardBuilder<>(transition, actions);
    Predicate<FsmContext<State, Event>> guard = ctx -> true;

    guardBuilder.guardedBy(guard);

    assertEquals(guard, transition.getGuard());
  }

  @Test
  void guardedTransition() throws InterruptedException {
    var fb = new FsmBuilder<State, Event>();

    var guardCondition = new AtomicBoolean(false);

    fb.when(State.S1)
        .on(Event.E1.class)
        .transitionTo(State.S2)
        .guardedBy(ctx -> guardCondition.get());

    var fsm = fb.build(State.S1);

    assertEquals(State.S1, fsm.fireEventBlocking(new Event.E1()));
    guardCondition.set(true);
    assertEquals(State.S2, fsm.fireEventBlocking(new Event.E1()));
  }

}

File: src/test/java/com/digitalpetri/strictmachine/dsl/GuardBuilderTest.kt
/*
 * Copyright 2018 Kevin Herron
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.digitalpetri.strictmachine.dsl

import com.digitalpetri.strictmachine.FsmContext
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.Predicate


class GuardBuilderTest {

    @Test
    fun `GuardBuilder#guardedBy() sets the guard condition on the transition`() {
        val transition = PredicatedTransition<State, Event>(
            { s -> s == State.S1 },
            { e -> e == Event.E1 },
            State.S2
        )
        val transitionActions = LinkedList<TransitionAction<State, Event>>()

        val gb = GuardBuilder(transition, transitionActions)

        val guard: Predicate<FsmContext<State, Event>> = Predicate { true }

        gb.guardedBy(guard)

        assertEquals(guard, transition.guard)
    }

    @Test
    fun `Transition only occurs when the guard condition is met`() {
        val fb = FsmBuilder<State, Event>()

        val guard = AtomicBoolean(false)

        fb.`when`(State.S1)
            .on(Event.E1)
            .transitionTo(State.S2)
            .guardedBy { guard.get() }

        val fsm = fb.build(State.S1)

        assertEquals(State.S1, fsm.fireEventBlocking(Event.E1))

        guard.set(true)

        assertEquals(State.S2, fsm.fireEventBlocking(Event.E1))
    }

}
